Subject: - Mellanox IB driver patch
From: Sven-Thorsten Dietrich <sdietrich@novell.com>
Date: Fri, 24 Aug 2007 02:25:26 -0700

Hi Ingo,

RT driver patch to eliminate in_atomic stack dump.

The problem code was identified by Michael S. Tsirkin, and he suggested
the fix.

I adapted to use RT's _nort primitives- should work correctly in all
configs.

Thanks,

Sven

Fixes in_atomic stack-dump, when Mellanox module
is loaded into the RT Kernel.

From: Michael S. Tsirkin <mst@dev.mellanox.co.il>

"Basically, if you just make spin_lock_irqsave (and spin_lock_irq) not disable
interrupts for non-raw spinlocks, I think all of infiniband will be fine without
changes."

signed-off-by: Sven-Thorsten Dietrich <sven@thebigcorporation.com>

Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 drivers/infiniband/ulp/ipoib/ipoib_multicast.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6-tip/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
===================================================================
--- linux-2.6-tip.orig/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
+++ linux-2.6-tip/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
@@ -773,7 +773,7 @@ void ipoib_mcast_restart_task(struct wor
 
 	ipoib_mcast_stop_thread(dev, 0);
 
-	local_irq_save(flags);
+	local_irq_save_nort(flags);
 	netif_addr_lock(dev);
 	spin_lock(&priv->lock);
 
@@ -852,7 +852,7 @@ void ipoib_mcast_restart_task(struct wor
 
 	spin_unlock(&priv->lock);
 	netif_addr_unlock(dev);
-	local_irq_restore(flags);
+	local_irq_restore_nort(flags);
 
 	/* We have to cancel outside of the spinlock */
 	list_for_each_entry_safe(mcast, tmcast, &remove_list, list) {
